File: crates/gui/frontend/src/hooks/basemaps.ts
/**
 * Offline-basemap commands + events (region store listing/deletion, download
 * planning, background downloads with `basemap:download` progress events,
 * viewport coverage checks) plus the pure formatting / visibility helpers
 * shared by the basemap UI.
 */

import { listen } from "@tauri-apps/api/event";
import { invoke, tryInvoke } from "./ipc";

// ── Types ──────────────────────────────────────────────────────────────────

/** Geographic bounding box as `[minLon, minLat, maxLon, maxLat]` (WGS84). */
export type BasemapBbox = [number, number, number, number];

/** Mirrors the backend `RegionInfo` DTO (list_basemap_regions). */
export interface BasemapRegionInfo {
  id: string;
  name: string;
  minLon: number;
  minLat: number;
  maxLon: number;
  maxLat: number;
  minZoom: number;
  maxZoom: number;
  /** Unix seconds at creation. */
  createdAt: number;
  /** Identifier of the planet build the tiles came from, when known. */
  planetBuild?: string | null;
  /** Bytes of tile data only this region claims. */
  uniqueBytes: number;
  /** Bytes of tile data shared with at least one other region. */
  sharedBytes: number;
  tileCount: number;
  /** Project IDs referencing this region. */
  projectIds: string[];
}

/** Mirrors the backend `BasemapStorageDto`. */
export interface BasemapStorage {
  regions: BasemapRegionInfo[];
  /** Actual bytes of the store on disk (db + WAL). */
  dataBytes: number;
  diskBytes: number;
  /** Regions no project references — safe to remove. */
  unusedRegionIds: string[];
}

/** Mirrors the backend `DownloadPlanDto` (plan_basemap_download). */
export interface BasemapDownloadPlan {
  /** Exact bytes a download would fetch. */
  newBytes: number;
  /** Bytes already on disk the region would share. */
  sharedBytes: number;
  missingTiles: number;
  presentTiles: number;
  /** In-bbox tiles the planet archive has no data for (open water). */
  absentTiles: number;
}

/** Mirrors the backend `CoverageDto` (basemap_coverage). */
export interface BasemapCoverage {
  presentTiles: number;
  totalTiles: number;
  covered: boolean;
}

export type BasemapDownloadPhase =
  | "planning"
  | "downloading"
  | "complete"
  | "cancelled"
  | "failed";

/** Payload of `basemap:download` events. */
export interface BasemapDownloadEvent {
  regionName: string;
  phase: BasemapDownloadPhase;
  doneBytes: number;
  totalBytes: number;
  /** Set on the `complete` event. */
  regionId?: string | null;
  /** Set on the `failed` event. */
  error?: string | null;
}

export const BASEMAP_DOWNLOAD_EVENT = "basemap:download";

// ── Commands ───────────────────────────────────────────────────────────────

/** Fetch the region store contents. Resolves `null` outside a Tauri shell
 *  (or on backend errors, which are reported via the shared IPC handler). */
export async function listBasemapRegions(): Promise<BasemapStorage | null> {
  return tryInvoke<BasemapStorage>("list_basemap_regions");
}

/** Resolve the exact download cost of a bbox without fetching tile data.
 *  SLOW — makes network range-requests against the planet archive. Rejects
 *  on backend errors so the modal can surface the failure. */
export async function planBasemapDownload(
  bbox: BasemapBbox,
  archiveUrl?: string,
): Promise<BasemapDownloadPlan> {
  return invoke<BasemapDownloadPlan>("plan_basemap_download", {
    bbox,
    archiveUrl: archiveUrl ?? null,
  });
}

/** Start a background region download. Progress, completion, and failure
 *  all arrive as `basemap:download` events; only one download runs at a
 *  time (a second call rejects). */
export async function downloadBasemapRegion(args: {
  name: string;
  bbox: BasemapBbox;
  projectId?: string | null;
  archiveUrl?: string | null;
}): Promise<void> {
  return invoke<void>("download_basemap_region", {
    name: args.name,
    bbox: args.bbox,
    projectId: args.projectId ?? null,
    archiveUrl: args.archiveUrl ?? null,
  });
}

/** Request cooperative cancellation of the in-flight download (if any).
 *  Confirmation arrives as a `cancelled` `basemap:download` event. */
export async function cancelBasemapDownload(): Promise<void> {
  await tryInvoke<void>("cancel_basemap_download");
}

/** Is the viewport bbox covered by stored tiles at (floored) `zoom`?
 *  Resolves `null` outside a Tauri shell. */
export async function getBasemapCoverage(
  bbox: BasemapBbox,
  zoom: number,
): Promise<BasemapCoverage | null> {
  return tryInvoke<BasemapCoverage>("basemap_coverage", { bbox, zoom });
}

/** Delete a stored region. Resolves the bytes actually freed (shared tiles
 *  survive). Rejects on backend errors so the caller can surface them. */
export async function deleteBasemapRegion(
  regionId: string,
): Promise<{ freedBytes: number }> {
  return invoke<{ freedBytes: number }>("delete_basemap_region", { regionId });
}

/** Reference `regionId` from `projectId` so it stops counting as unused. */
export async function linkProjectBasemapRegion(
  projectId: string,
  regionId: string,
): Promise<void> {
  await tryInvoke<void>("link_project_basemap_region", {
    projectId,
    regionId,
  });
}

/** Subscribe to `basemap:download` progress events from the backend.
 *  Returns the unlisten function — call it to unsubscribe. */
export function listenBasemapDownload(
  cb: (e: BasemapDownloadEvent) => void,
): Promise<() => void> {
  return listen<BasemapDownloadEvent>(BASEMAP_DOWNLOAD_EVENT, (ev) =>
    cb(ev.payload),
  );
}

// ── Pure helpers (unit-tested in basemaps.test.ts) ─────────────────────────

/** Human-readable byte size: "0 B", "512 B", "1.5 KB", "12.3 MB", "1.2 GB".
 *  Binary (1024) steps; one decimal, with a trailing ".0" trimmed. */
export function formatBytes(bytes: number): string {
  if (!Number.isFinite(bytes) || bytes < 0) return "0 B";
  if (bytes < 1024) return `${Math.round(bytes)} B`;
  const units = ["KB", "MB", "GB", "TB"];
  let value = bytes;
  let unit = "B";
  for (const u of units) {
    if (value < 1024) break;
    value /= 1024;
    unit = u;
  }
  const rounded = value.toFixed(1).replace(/\.0$/, "");
  return `${rounded} ${unit}`;
}

/** Region size summary: `"12.3 MB unique · 4.1 MB shared"`. */
export function regionSizeLabel(r: {
  uniqueBytes: number;
  sharedBytes: number;
}): string {
  return `${formatBytes(r.uniqueBytes)} unique · ${formatBytes(r.sharedBytes)} shared`;
}

/** Web-Mercator latitude limit — tiles don't exist beyond it. */
const MAX_MERCATOR_LAT = 85.051129;

/** Grow `bbox` by `fraction` of its span on every side, clamped to the
 *  world (lon ±180, lat ±85.05 Web-Mercator limit). */
export function padBbox(bbox: BasemapBbox, fraction: number): BasemapBbox {
  const [minLon, minLat, maxLon, maxLat] = bbox;
  const lonPad = (maxLon - minLon) * fraction;
  const latPad = (maxLat - minLat) * fraction;
  return [
    Math.max(-180, minLon - lonPad),
    Math.max(-MAX_MERCATOR_LAT, minLat - latPad),
    Math.min(180, maxLon + lonPad),
    Math.min(MAX_MERCATOR_LAT, maxLat + latPad),
  ];
}

/** Parse the download modal's four bbox fields. Returns `null` unless all
 *  four parse as finite numbers within world bounds with min < max. */
export function bboxFromStrings(
  minLon: string,
  minLat: string,
  maxLon: string,
  maxLat: string,
): BasemapBbox | null {
  const nums = [minLon, minLat, maxLon, maxLat].map((s) =>
    Number.parseFloat(s.trim()),
  );
  if (nums.some((n) => !Number.isFinite(n))) return null;
  const [w, s, e, n] = nums;
  if (w < -180 || e > 180 || s < -90 || n > 90) return null;
  if (w >= e || s >= n) return null;
  return [w, s, e, n];
}

/** True for the locally-served basemap styles ("offline-*"). */
export function isOfflineBasemap(basemap: string): boolean {
  return basemap.startsWith("offline-");
}

/** Region downloads only store street-detail zooms ≥ 7; below that the
 *  coverage chip is meaningless (the world overview covers 0–6). */
export const COVERAGE_MIN_ZOOM = 7;

/** Should the "No offline detail here" chip render? Pure so the visibility
 *  matrix is unit-testable. */
export function shouldShowCoverageChip(args: {
  /** Selected basemap style value — chip is offline-only. */
  basemap: string;
  /** Current canvas view mode — chip is map-mode only. */
  viewMode: string;
  /** Current map zoom, or null before the first move-end report. */
  zoom: number | null;
  /** Coverage of the current viewport, or null while unknown/stale. */
  covered: boolean | null;
  /** True while a region download is in flight. */
  downloadActive: boolean;
}): boolean {
  return (
    isOfflineBasemap(args.basemap) &&
    args.viewMode === "map" &&
    args.zoom != null &&
    args.zoom >= COVERAGE_MIN_ZOOM &&
    args.covered === false &&
    !args.downloadActive
  );
}

File: crates/gui/src/commands/basemaps.rs
//! Offline-basemap commands: region listing/deletion, download planning
//! and execution with progress events, viewport coverage checks, and
//! project↔region links.
//!
//! Command handlers stay thin — everything stateful lives in
//! [`crate::basemap`] — so the logic under them is unit-tested without a
//! Tauri runtime. Downloads run on a worker thread (blocking range
//! requests) and report through `basemap:download` events, mirroring the
//! run-queue event pattern.

use std::sync::atomic::AtomicBool;
use std::sync::Arc;

use serde::Serialize;
use tauri::Emitter;

use crate::basemap::{
    plan_extract, run_extract, tiles_in_bbox, Archive, BasemapState, HttpSource, RegionInfo,
    TileCoord,
};

/// Protomaps publishes daily planet builds; this pinned build is the
/// default source and can be overridden per download call.
const DEFAULT_PLANET_URL: &str = "https://build.protomaps.com/20260725.pmtiles";
/// Street-detail zoom window for region downloads (the world overview
/// region covers 0–6 when the user opts into it).
const REGION_MIN_ZOOM: u8 = 7;
const REGION_MAX_ZOOM: u8 = 15;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BasemapStorageDto {
    pub regions: Vec<RegionInfo>,
    /// Bytes of downloaded tile data.
    pub data_bytes: u64,
    /// Actual bytes of the store on disk (db + WAL, incl. container overhead).
    pub disk_bytes: u64,
    pub unused_region_ids: Vec<String>,
}

#[tauri::command(async)]
pub fn list_basemap_regions(
    state: tauri::State<'_, BasemapState>,
) -> Result<BasemapStorageDto, String> {
    let store = state.store()?;
    Ok(BasemapStorageDto {
        regions: store.list_regions()?,
        data_bytes: store.data_bytes()?,
        disk_bytes: store.disk_bytes(),
        unused_region_ids: store.unused_regions()?,
    })
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FreedDto {
    pub freed_bytes: u64,
}

#[tauri::command(async)]
pub fn delete_basemap_region(
    state: tauri::State<'_, BasemapState>,
    region_id: String,
) -> Result<FreedDto, String> {
    Ok(FreedDto {
        freed_bytes: state.store()?.delete_region(&region_id)?,
    })
}

#[tauri::command(async)]
pub fn link_project_basemap_region(
    state: tauri::State<'_, BasemapState>,
    project_id: String,
    region_id: String,
) -> Result<(), String> {
    state.store()?.link_project(&project_id, &region_id)
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DownloadPlanDto {
    /// Exact bytes a download would fetch.
    pub new_bytes: u64,
    /// Bytes already on disk the region would share.
    pub shared_bytes: u64,
    pub missing_tiles: u64,
    pub present_tiles: u64,
    /// In-bbox tiles the planet archive has no data for (open water).
    pub absent_tiles: u64,
}

/// Resolve the exact download cost of a bbox without fetching tile data.
#[tauri::command(async)]
pub fn plan_basemap_download(
    state: tauri::State<'_, BasemapState>,
    bbox: (f64, f64, f64, f64),
    archive_url: Option<String>,
) -> Result<DownloadPlanDto, String> {
    let store = state.store()?;
    let url = archive_url.as_deref().unwrap_or(DEFAULT_PLANET_URL);
    let mut archive = Archive::open(HttpSource::new(url)?)?;
    let plan = plan_extract(store, &mut archive, bbox, REGION_MIN_ZOOM, REGION_MAX_ZOOM)?;
    Ok(DownloadPlanDto {
        new_bytes: plan.new_bytes,
        shared_bytes: plan.shared_bytes,
        missing_tiles: plan.missing.len() as u64,
        present_tiles: plan.present.len() as u64,
        absent_tiles: plan.absent,
    })
}

/// Progress payload for `basemap:download` events.
#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
struct DownloadEvent {
    region_name: String,
    phase: &'static str,
    done_bytes: u64,
    total_bytes: u64,
    region_id: Option<String>,
    error: Option<String>,
}

/// Start a region download on a worker thread. Progress, completion, and
/// failure all arrive as `basemap:download` events; only one download
/// runs at a time.
#[tauri::command]
pub fn download_basemap_region(
    app: tauri::AppHandle,
    state: tauri::State<'_, BasemapState>,
    name: String,
    bbox: (f64, f64, f64, f64),
    project_id: Option<String>,
    archive_url: Option<String>,
) -> Result<(), String> {
    let cancel = Arc::new(AtomicBool::new(false));
    state.begin_download(cancel.clone())?;
    let url = archive_url.unwrap_or_else(|| DEFAULT_PLANET_URL.to_string());
    let planet_build = url
        .rsplit('/')
        .next()
        .and_then(|f| f.strip_suffix(".pmtiles"))
        .map(str::to_string);

    std::thread::spawn(move || {
        use tauri::Manager;
        let state = app.state::<BasemapState>();
        let emit = |payload: DownloadEvent| {
            if let Err(e) = app.emit("basemap:download", payload) {
                tracing::warn!("basemap download event emit failed: {e}");
            }
        };
        let event = |phase: &'static str, done: u64, total: u64| DownloadEvent {
            region_name: name.clone(),
            phase,
            done_bytes: done,
            total_bytes: total,
            region_id: None,
            error: None,
        };

        let result = (|| -> Result<(String, u64), String> {
            let store = state.store()?;
            let mut archive = Archive::open(HttpSource::new(&url)?)?;
            emit(event("planning", 0, 0));
            let plan = plan_extract(store, &mut archive, bbox, REGION_MIN_ZOOM, REGION_MAX_ZOOM)?;
            let total = plan.new_bytes;
            let region_id = run_extract(
                store,
                &mut archive,
                &plan,
                &name,
                planet_build.as_deref(),
                &mut |done, total| emit(event("downloading", done, total)),
                &cancel,
            )?;
            if let Some(project) = &project_id {
                store.link_project(project, &region_id)?;
            }
            Ok((region_id, total))
        })();

        state.end_download();
        match result {
            Ok((region_id, total)) => emit(DownloadEvent {
                region_name: name.clone(),
                phase: "complete",
                done_bytes: total,
                total_bytes: total,
                region_id: Some(region_id),
                error: None,
            }),
            Err(e) => emit(DownloadEvent {
                region_name: name.clone(),
                phase: if e == "cancelled" {
                    "cancelled"
                } else {
                    "failed"
                },
                done_bytes: 0,
                total_bytes: 0,
                region_id: None,
                error: (e != "cancelled").then_some(e),
            }),
        }
    });
    Ok(())
}

#[tauri::command]
pub fn cancel_basemap_download(state: tauri::State<'_, BasemapState>) {
    state.cancel_download();
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CoverageDto {
    /// Present / total tiles for the viewport at the checked zoom.
    pub present_tiles: u64,
    pub total_tiles: u64,
    pub covered: bool,
}

/// Is the viewport covered by stored tiles at (clamped) `zoom`?
#[tauri::command(async)]
pub fn basemap_coverage(
    state: tauri::State<'_, BasemapState>,
    bbox: (f64, f64, f64, f64),
    zoom: f64,
) -> Result<CoverageDto, String> {
    let z = (zoom.floor().clamp(0.0, f64::from(REGION_MAX_ZOOM))) as u8;
    let coords: Vec<TileCoord> = tiles_in_bbox(bbox, z);
    let total = coords.len() as u64;
    let present = state.store()?.count_present(&coords)?;
    Ok(CoverageDto {
        present_tiles: present,
        total_tiles: total,
        covered: total > 0 && present == total,
    })
}

/// Project-deletion hook: release the project's region references so
/// affected regions surface as unused rather than lingering silently.
pub(crate) fn release_project_regions(app: &tauri::AppHandle, project_id: &str) {
    use tauri::Manager;
    let state = app.state::<BasemapState>();
    match state.store() {
        Ok(store) => {
            if let Err(e) = store.unlink_project(project_id) {
                tracing::warn!("failed to release basemap regions of {project_id}: {e}");
            }
        }
        Err(e) => tracing::warn!("basemap store unavailable during project delete: {e}"),
    }
}
